### PR TITLE
feat: use cluster identity for using azcopy in volume cloning

### PR DESCRIPTION
**Pull request by umagnus**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

azcopy already supports managed identity in a different way: [Authorize access to blobs with AzCopy & Microsoft Entra ID | Microsoft Learn](https://learn.microsoft.com/en-us/azure/storage/common/storage-use-azcopy-authorize-azure-active-directory#authorize-a-managed-identity), and it supports file in the latest version.
if user only provides storage account key, then we should use sas token; otherwise we should leverage aks cluster identity(managed identity or spn) for volume cloning which is better way.
**Which issue(s) this PR fixes**:

Fixes #

**Requirements**:
- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

**Release note**:
```
none
```

### CODE REVIEW

File: pkg/azurefile/azurefile.go [hunk 1]
```diff
@@ -1014,15 +1022,44 @@ func (d *Driver) copyFileShare(req *csi.CreateVolumeRequest, accountKey string,
...
+				if len(authAzcopyEnv) > 0 {
+					cmd.Env = append(os.Environ(), authAzcopyEnv...)
+				}
+				out, copyErr := cmd.CombinedOutput()
 				if copyErr != nil {
+					if strings.Contains(string(out), authorizationPermissionMismatch) {
+						klog.Warningf("azcopy list failed with AuthorizationPermissionMismatch error, should assign \"Storage File Data SMB Share Elevated Contributor\" role to controller identity, fall back to use sas token, original output: %v", string(out))
+						d.azcopySasTokenCache.Set(accountName, "")
+						var accountKey, sasToken string
+						if accountKey, err = d.GetStorageAccesskey(ctx, accountOptions, secrets, secretName, secretNamespace); err != nil {
```
Comment (andyzhangx) [inline pkg/azurefile/azurefile.go:1042]: you could leverage `getSASToken` func to get sas token? thus don't need to duplicate code

File: pkg/azurefile/azurefile.go [hunk 2]
```diff
@@ -1014,15 +1022,44 @@ func (d *Driver) copyFileShare(req *csi.CreateVolumeRequest, accountKey string,
...
+							return err
+						}
+						klog.V(2).Infof("generate sas token for account(%s) and fallback to copy fileshare %s to %s", accountName, srcFileShareName, dstFileShareName)
+						if sasToken, err = generateSASToken(accountName, accountKey, storageEndpointSuffix, d.sasTokenExpirationMinutes); err != nil {
+							return err
+						}
+						srcPath := srcPath + sasToken
+						dstPath := dstPath + sasToken
+						cmd := exec.Command("azcopy", "copy", srcPath, dstPath, "--recursive", "--check-length=false")
+						if len(authAzcopyEnv) > 0 {
```
Comment (andyzhangx) [inline pkg/azurefile/azurefile.go:1052]: remove this since it's already using sastoken

File: pkg/azurefile/azurefile.go [hunk 3]
```diff
@@ -1014,15 +1022,44 @@ func (d *Driver) copyFileShare(req *csi.CreateVolumeRequest, accountKey string,
...
 			case fileutil.AzcopyJobNotFound:
 				klog.V(2).Infof("copy fileshare %s to %s", srcFileShareName, dstFileShareName)
-				out, copyErr := exec.Command("azcopy", "copy", srcPath, dstPath, "--recursive", "--check-length=false").CombinedOutput()
+				cmd := exec.Command("azcopy", "copy", srcPath, dstPath, "--recursive", "--check-length=false")
+				if len(authAzcopyEnv) > 0 {
+					cmd.Env = append(os.Environ(), authAzcopyEnv...)
+				}
+				out, copyErr := cmd.CombinedOutput()
 				if copyErr != nil {
+					if strings.Contains(string(out), authorizationPermissionMismatch) {
```
Comment (andyzhangx) [inline pkg/azurefile/azurefile.go:1038]:
```
if len(accountSASToken) > 0 && strings.Contains(string(out), authorizationPermissionMismatch) {

```

File: pkg/azurefile/azurefile.go [hunk 4]
```diff
@@ -1014,15 +1022,44 @@ func (d *Driver) copyFileShare(req *csi.CreateVolumeRequest, accountKey string,
...
+						d.azcopySasTokenCache.Set(accountName, "")
+						var accountKey, sasToken string
+						if accountKey, err = d.GetStorageAccesskey(ctx, accountOptions, secrets, secretName, secretNamespace); err != nil {
+							return err
+						}
+						klog.V(2).Infof("generate sas token for account(%s) and fallback to copy fileshare %s to %s", accountName, srcFileShareName, dstFileShareName)
+						if sasToken, err = generateSASToken(accountName, accountKey, storageEndpointSuffix, d.sasTokenExpirationMinutes); err != nil {
+							return err
+						}
+						srcPath := srcPath + sasToken
```
Comment (andyzhangx) [inline pkg/azurefile/azurefile.go:1049]: this is a loop, don't change `srcPath`, `dstPath`, otherwise this append may happen multiple times

**Review by andyzhangx — COMMENTED** [review]
pls also follow this PR to refine getSASToken func:

https://github.com/kubernetes-sigs/blob-csi-driver/pull/1207

File: pkg/azurefile/controllerserver_test.go [hunk 5]
```diff
@@ -2904,3 +2936,256 @@ func TestGenerateSASToken(t *testing.T) {
 		})
 	}
 }
+
+func Test_authorizeAzcopyWithIdentity(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (umagnus) [reply pkg/azurefile/controllerserver_test.go]: done

File: pkg/azurefile/azurefile.go (hunk 1 above)
Comment (umagnus) [reply pkg/azurefile/azurefile.go]: done, give a useSasToken parameter to get sas token

File: pkg/azurefile/azurefile.go (hunk 2 above)
Comment (umagnus) [reply pkg/azurefile/azurefile.go]: done

File: pkg/azurefile/azurefile.go (hunk 3 above)
Comment (umagnus) [reply pkg/azurefile/azurefile.go]: I think it should be `accountSASToken ==""` which means we use identity and get permission mismatch

File: pkg/azurefile/azurefile.go (hunk 4 above)
Comment (umagnus) [reply pkg/azurefile/azurefile.go]: done

**Comment by andyzhangx** [comment]
can you squash all commits? thanks.

File: pkg/azurefile/controllerserver.go [hunk 6]
```diff
@@ -1300,6 +1309,77 @@ func isValidVolumeCapabilities(volCaps []*csi.VolumeCapability) error {
...
+			return []string{}, fmt.Errorf("AADClientID and TenantID must be set when use service principal")
+		}
+		authAzcopyEnv = append(authAzcopyEnv, fmt.Sprintf("%s=%s", azcopySPAApplicationID, azureAuthConfig.AADClientID))
+		authAzcopyEnv = append(authAzcopyEnv, fmt.Sprintf("%s=%s", azcopySPAClientSecret, azureAuthConfig.AADClientSecret))
+		authAzcopyEnv = append(authAzcopyEnv, fmt.Sprintf("%s=%s", azcopyTenantID, azureAuthConfig.TenantID))
+		klog.V(2).Infof(fmt.Sprintf("set AZCOPY_SPA_APPLICATION_ID=%s, AZCOPY_TENANT_ID=%s successfully", azureAuthConfig.AADClientID, azureAuthConfig.TenantID))
+
+		return authAzcopyEnv, nil
+	}
+	return []string{}, fmt.Errorf("service principle or managed identity are both not set")
```
[the remark this answers is not in the document]
Comment (umagnus) [reply pkg/azurefile/controllerserver.go]: done

File: pkg/azurefile/controllerserver.go [hunk 7]
```diff
@@ -1300,6 +1309,77 @@ func isValidVolumeCapabilities(volCaps []*csi.VolumeCapability) error {
...
+}
+
+// getAzcopyAuth will only generate sas token for azcopy in following conditions:
+// 1. secrets is not empty
+// 2. driver is not using managed identity and service principal
+// 3. azcopy returns AuthorizationPermissionMismatch error when using service principal or managed identity
+// 4. parameter useSasToken is true
+func (d *Driver) getAzcopyAuth(ctx context.Context, accountName, accountKey, storageEndpointSuffix string, accountOptions *azure.AccountOptions, secrets map[string]string, secretName, secretNamespace string, useSasToken bool) (string, []string, error) {
+	var authAzcopyEnv []string
+	if len(secrets) == 0 && len(secretName) == 0 {
```
[the remark this answers is not in the document]
Comment (umagnus) [reply pkg/azurefile/controllerserver.go]: done

File: pkg/azurefile/azurefile.go [hunk 8]
```diff
@@ -1014,14 +1015,27 @@ func (d *Driver) copyFileShare(req *csi.CreateVolumeRequest, accountKey string,
...
-			jobState, percent, err := d.azcopy.GetAzcopyJob(dstFileShareName)
+			jobState, percent, err := d.azcopy.GetAzcopyJob(dstFileShareName, authAzcopyEnv)
 			klog.V(2).Infof("azcopy job status: %s, copy percent: %s%%, error: %v", jobState, percent, err)
 			switch jobState {
 			case fileutil.AzcopyJobError, fileutil.AzcopyJobCompleted:
 				return err
 			case fileutil.AzcopyJobNotFound:
 				klog.V(2).Infof("copy fileshare %s to %s", srcFileShareName, dstFileShareName)
-				out, copyErr := exec.Command("azcopy", "copy", srcPath, dstPath, "--recursive", "--check-length=false").CombinedOutput()
+				cmd := exec.Command("azcopy", "copy", srcPath, dstPath, "--recursive", "--check-length=false")
```
[the remark this answers is not in the document]
Comment (umagnus) [reply pkg/azurefile/azurefile.go]: done

File: pkg/azurefile/azurefile.go [hunk 9]
```diff
@@ -1014,14 +1015,27 @@ func (d *Driver) copyFileShare(req *csi.CreateVolumeRequest, accountKey string,
...
+				}
+				out, copyErr := cmd.CombinedOutput()
+				if accountSASToken == "" && strings.Contains(string(out), authorizationPermissionMismatch) && copyErr != nil {
+					klog.Warningf("azcopy list failed with AuthorizationPermissionMismatch error, should assign \"Storage File Data SMB Share Elevated Contributor\" role to controller identity, fall back to use sas token, original output: %v", string(out))
+					d.azcopySasTokenCache.Set(accountName, "")
+					var sasToken string
+					if sasToken, _, err = d.getAzcopyAuth(ctx, accountName, "", storageEndpointSuffix, accountOptions, secrets, secretName, secretNamespace, true); err != nil {
+						return err
+					}
+					out, copyErr = exec.Command("azcopy", "copy", srcPath+sasToken, dstPath+sasToken, "--recursive", "--check-length=false").CombinedOutput()
```
[the remark this answers is not in the document]
Comment (umagnus) [reply pkg/azurefile/azurefile.go]: done

**Pull request merged**